feat: Add app version checker plugin

gorandalum opened the pull request:
If a MIN_APP_VERSION env variable is set, then all incoming requests
with app version in http header (atb-app-version) lower than the
minimum is rejected with status code 426 Upgrade Required.

marius-at-atb reviewed (approved):
Lgtm!
Since `compare-version.ts` is a copy of the one from the app, should there be comments both in this file and the one in the app to keep them in sync if making changes?
Also since `compare-version.test.ts` is an extract from `utils.test.ts` in the app, I think this should be extracted in the app as well, so that the files are identically matching.

rosvik commented:
We use some bff endpoints in the webshop as well, will those break after this change? https://github.com/AtB-AS/webshop2/blob/main/src/server-api-service/journey/index.ts

gorandalum commented:
> We use some bff endpoints in the webshop as well, will those break after this change? https://github.com/AtB-AS/webshop2/blob/main/src/server-api-service/journey/index.ts
> 
> Edit: I guess it's ok as long as `appVersion` isn't set.

Yeah, they won't break, but you poke into something that need to be handled:
https://github.com/AtB-AS/kundevendt/issues/19399#issuecomment-2448399057

gorandalum commented:
I added a check on install id to determine if the request is coming from the app. The install id is sent from all released app versions.

gorandalum commented:
> I added a check on install id to determine if the request is coming from the app. The install id is sent from all released app versions.

… and now changed the logic to use the new atb-webshop-version introduced for the webshop.

gorandalum commented:
Added you as reviewer as well @mjenssen, to give input on what error we should return in this situation, so it is the same across our services.

mjenssen commented on src/plugins/app-version-checker-plugin.ts:
```diff
@@ -0,0 +1,35 @@
+      if (isAppVersionTooLow(request, minAppVersion)) {
+        return h
+          .response({error: `Required client app version: ${minAppVersion}`})
+          .code(426)
```
We shouldn't return 426 here, as it indicates that the HTTP _protocol_ needs to be upgraded, ie. if the client uses HTTP 1.1 but the server requires 2.0. It is also required that you respond with an `Upgrade` header saying which protocols are allowed. See [RFC 7231, section 6.5.15](https://datatracker.ietf.org/doc/html/rfc7231#section-6.5.15) for more details.

We could go for 406 (Not Acceptable), as it seems to indicate that it's not acceptable for the user-agent (which is what we are doing). See [RFC 7231, setion 6.5.6](https://datatracker.ietf.org/doc/html/rfc7231#section-6.5.6).

gorandalum commented:
> Lgtm! Since `compare-version.ts` is a copy of the one from the app, should there be comments both in this file and the one in the app to keep them in sync if making changes? Also since `compare-version.test.ts` is an extract from `utils.test.ts` in the app, I think this should be extracted in the app as well, so that the files are identically matching.

One of the benefits of copying a simple utility function is that it is done quickly, and doesn't impose any dependencies or connections between the two implementations/code-blocks. You just "copy-and-forget!". I think this is the logical thing for many utility functions which probably won't change at any time, for example the `isDefined` function we have in the app and webshop. 

I ended up doing the "copy and forget!" for the `compareVersion` as well, however I see that it is quite a large utility function, and also of such complexity that it has many unit-test. It kinda feels so big that it maybe should have been shared between app/webshop and bff, and I think it is a good candidate for our first shared lib function. If anyone are up for it, then just do it, create a shared lib!
